### PR TITLE
s4627365 -- Siamese Network

### PR DESCRIPTION
**Pull request by JasonCodeMaker**
This pull request is for the final submission of the COMP3710 Report.
The target of the project is to classify Alzheimer’s disease of ADNI brain dataset using a classifier based on the Siamese network

The project contains  4 python files and a README.md
  "dataset.py":  containing the data loader for loading and preprocessing the dataset
  "modules.py": containing the source code of the components of the model
  "train.py":  containing the source code for training, validating, testing and saving the model. 
  "predict.py": showing example usage of the trained model. Visualizing some results.  
  README.md: describe the structure of the Siamese Network, the architecture and the performance of my modules

Zecheng Zhao
s4627365

### CODE REVIEW

**Comment by nathasha-naranpanawa** [comment]
- Commit messages are OK
- Readme is good.
- Would have been better if loss and accuracy were plotted separately
- Not sure if 2D or 3D data were used. Not mentioned 
- Performance is good (although the model seems to be overfitting)
- Loss function should be a contrastive loss for the Siamese. Instead, a classification loss  is used for the similar/non-similar classification (slightly incorrect)
- Classifier architecture and predictions with classifier not clearly explained. The classification should be for embeddings of individual images

**Comment by JasonCodeMaker** [comment]
Sorry for not explaining clearly in the README, I mensioned the input shape of the image to be (128*128*1) so each input is 2D with 1 chanel.
For the final prediction, as I mensioned in the prediction part of README,  I constructed another neural network classifier. This classifier will take images (128*128*1) as input and it will use the trained embedding model to convert the image to a 256 1d vector. Then it use a output layer with softmax activation function to classify the image, which can classify the individual images.

**Comment by shakes76** [comment]
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate use and implementation (loss function noted above) -2
Good spacing and comments
Header blocks
## Recognition Problem 
Solves problem
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage 
Module present
Commenting
No Data leakage
Difficulty: Normal -5
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe acceptable, couldve had more explanation -1
Good Description and Comments 
Markdown used PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
